lib/Conversion/PolyToStandard/PolyToStandard.cpp

```diff
@@ -0,0 +1,156 @@
+#include "include/Conversion/PolyToStandard/PolyToStandard.h"
+
+#include "include/Dialect/Poly/IR/PolyOps.h"
+#include "include/Dialect/Poly/IR/PolyTypes.h"
+#include "mlir/include/mlir/Dialect/Arith/IR/Arith.h"  // from @llvm-project
+#include "mlir/include/mlir/Dialect/Func/Transforms/FuncConversions.h"  // from @llvm-project
+#include "mlir/include/mlir/Dialect/Linalg/IR/Linalg.h"  // from @llvm-project
+#include "mlir/include/mlir/Dialect/Tensor/IR/Tensor.h"  // from @llvm-project
+#include "mlir/include/mlir/Dialect/Utils/StructuredOpsUtils.h"  // from @llvm-project
+#include "mlir/include/mlir/IR/PatternMatch.h"  // from @llvm-project
+#include "mlir/include/mlir/Transforms/DialectConversion.h"  // from @llvm-project
+
+namespace mlir {
+namespace heir {
+namespace poly {
+
+#define GEN_PASS_DEF_POLYTOSTANDARD
+#include "include/Conversion/PolyToStandard/PolyToStandard.h.inc"
+
+static Value materializeSource(OpBuilder &builder, Type type, ValueRange inputs,
+                               Location loc) {
+  assert(inputs.size() == 1);
+  auto inputType = inputs[0].getType();
+  if (isa<PolynomialType>(inputType))
+    llvm_unreachable(
+        "A polynomial should never be the input to a materializeSource.");
+
+  if (!isa<TensorType>(inputType)) {
+    llvm::errs() << "materializeSource should only try to materialize tensor "
+                    "types back to "
+                    "polynomials, but found type: "
+                 << inputType;
+    llvm_unreachable("");
+  }
+
+  PolynomialType targetType = dyn_cast<PolynomialType>(inputType);
+  if (!targetType) llvm_unreachable("Target type must be a poly.poly");
+
+  return builder.create<PolyFromCoeffsOp>(loc, inputs[0], targetType.getRing());
+}
+
+static Value materializeTarget(OpBuilder &builder, Type type, ValueRange inputs,
+                               Location loc) {
+  assert(inputs.size() == 1);
+  auto inputType = inputs[0].getType();
+  if (!isa<PolynomialType>(inputType)) {
+    llvm::errs() << "materializeTarget should only try to materialize "
+                    "polynomial types to "
+                    "legal types, but found input: "
+                 << inputType;
+    llvm_unreachable("");
+  }
+
+  auto polyType = dyn_cast<PolynomialType>(inputType);
+  uint32_t idealDegree = polyType.getRing().ideal().getDegree();
+  auto startOp =
+      builder.create<arith::ConstantIntOp>(loc, 0, builder.getIndexType());
+  auto endOp = builder.create<arith::ConstantIntOp>(loc, idealDegree,
+                                                    builder.getIndexType());
+  return builder.create<ExtractSliceOp>(loc, inputs[0], startOp.getResult(),
+                                        endOp.getResult());
```

Comment:
As mentioned above, I'll remove the materializations anyway, so that will be unrelated to this PR.

But while I'm not opposed to adding poly.to_coeffs, I guess I'm not clear on the best practices around having ops that are specializations or other ops. Wouldn't we want to canonicalize to_coeffs as an extract_slice, and write all the patterns in terms of extract_slice as the most general case? Or have additional patterns for both of the ops separately, so that it's easier to write patterns on `to_coeffs` as you mentioned.